File: 백준/Gold/11054. 가장 긴 바이토닉 부분 수열/가장 긴 바이토닉 부분 수열.py

from sys import stdin

input = stdin.readline


def LDS(array, n):
    dp = [1] * n
    for i in range(n - 2, -1, -1):
        for k in range(n - 1, i, -1):
            if array[k] < array[i]:
                dp[i] = max(dp[i], dp[k] + 1)
    return dp


def LIS(array, n):
    dp = [1] * n
    for i in range(n):
        for k in range(i):
            if array[k] < array[i]:
                dp[i] = max(dp[i], dp[k] + 1)
    return dp


def bitonic_sequence(n, array):

    dp_inc = LIS(array, n)
    dp_dec = LDS(array, n)
    answer = [dp_inc[i] + dp_dec[i] - 1 for i in range(n)]

    return max(answer)


if __name__ == "__main__":
    N = int(input())

    S = list(map(int, input().split()))
    """
    10
    1 5 2 1 4 3 4 5 2 1
    7
    (1,2,3,4,5,2,1)이렇게 길이가 7인 부분수열이 답.
    """

    print(bitonic_sequence(N, S))

File: 백준/Gold/11054. 가장 긴 바이토닉 부분 수열/가장 긴 바이토닉 부분 수열.java
import java.util.*;

public class Main {


    public static int[] LDS(ArrayList<Integer> array, int n){
        int [] dp = new int[n];
        Arrays.fill(dp,1);
        for(int i=n-2; i>-1; i--){
            for (int k=n-1; k>i; k--){
                if (array.get(k) < array.get(i)){
                    dp[i] = Math.max(dp[i],dp[k] +1);
                }
            }
        }
        return dp;
    }

    public static int[] LIS(ArrayList<Integer> array, int n){
        int[] dp = new int[n];
        Arrays.fill(dp,1);

        for(int i=0; i<n; i++){
            for (int k=0; k<i; k++){
                if(array.get(k) < array.get(i)){
                    dp[i] = Math.max(dp[i],dp[k]+1);
                }
            }
        }

        return dp;
    }
    public static int solution(ArrayList<Integer> array, int n){
        ArrayList<Integer> answer = new ArrayList<>();
        int [] dp_inc = LIS(array,n);
        int [] dp_dec = LDS(array,n);

        for (int i=0; i<n; i++){
            answer.add(dp_inc[i] + dp_dec[i] - 1);
        }
        return Collections.max(answer);
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int N = sc.nextInt();
        ArrayList<Integer> array = new ArrayList<>();
        for (int i=0; i<N; i++){
            array.add(sc.nextInt());
        }

        System.out.println(solution(array,N));
        sc.close();
    }
}
